Поменял структуру программы

Gaileks opened the pull request.

Gaileks committed:
Поменял структуру программы

danelyan commented on Сoursework/LearnEnglishWordsTelegramBot/src/main/kotlin/ProgramLogic.kt:
```diff
@@ -0,0 +1,73 @@
+
+class ProgramLogic {
+
+    private fun getNotLearnedWord(): MutableList<Word> {
```
Для чего возвращать мутабельный список? Старайся использовать минимально конкретные классы.
Иначе если ты где то дальше пропишешь MutableList (хоть он там и не нужен) уже отказаться не получится.

danelyan reviewed (changes requested):
Как гласит древняя мудрость: "Лучший код тот, который не написан".
Ненаписанный код не сожержит багов, не будет падать, его не нужно покрывать тестами, не нужно рефакторить, его даже читать не нужно.

Возьмем для примера класс WordFile:

```Kotlin
class WordFile {
    private val wordFile = File("words.txt")
    val linesFile = wordFile.readLines()

    fun writeTextToFile(textToWriting: String) {
        wordFile.writeText(textToWriting)
    }
}
```

- читает файл с именем `words.txt` (только с этим именем)
- текст читается в переменную сразу при создании класса
- пишет текст в файл words.txt (и только в него)

Как мы его применяем? Давай прочитаем строки и запишем:
```Kotlin
val wf = WordFile()
val lines = wf.linesFile()

// тут что-то делаем, например склеим, не важно
val all = lines.join()

wf.writeTextToFile(all)
```

А теперь без этого класса:

```Kotlin
val wf = File("words.txt")
val lines = wf.readLines()

// тут что-то делаем, например склеим, не важно
val all = lines.join()

wf.writeText(all)
```

Кода столько же, но твой вариант читается хуже. 
Ведь `File`, `readLines`, `writeText` - штуки знакомые.
А твои конструкции нужно еще проверить.

UploadSaveData - полезный класс, служит для загрузки словаря. 
Только я бы его тогда назвал его DictionaryDataStore - чтобы подчеркнуть, что это класс для загрузки/хранения словаря. Или даже FileDictionaryDataStore - ведь он работает с файловым словарем. В будущем можно было сделать интерфейс DictionaryDataStore, а для него реализации FileDictionaryDataStore, DatabaseDictionaryDataStore, ApiDictionaryDataStore...

ProgramLogic - неплохой класс, содержит в себе логику работы тренажера. Но он не универсальный. Ты не сможешь просто взять его и переиспользовать в телеграм боте. 
А при каждом вызове почти любого метода ты конструируешь UploadSaveData(). Что если внутри этого класса нужно подключаться к базе данных? Каждый раз будешь устанавливать соединение?
Я к тому что этот класс не хранит внутри себя никакого состояния, никаких ссылок на внешние полезные классы. Ну и название никак не отражает что это класс для изучения слов.

Statistics - плохой класс, он "неявно" при создании класса грузит словарь (причем дважды - для каждой переменной). Это делает его не переиспользуемым после создания. Ведь он посчитал статистику и все. Я бы оставил его простым data классом, хранящим подсчитанные данные. А сами данные подситывал бы внутри ProgramLogic. Ведь подсчет статистики это логика программы.

Круто что ты экспериментируешь! 
Попробуй посмотреть теперь на свои классы под вышеописанным углом и давай сократим их количество.

Слишком много классов для простого тренажера :)

<img width="400" alt="Screenshot 2023-09-27 at 22 47 08" src="https://github.com/Gaileks/LearnEnglishWordsTelegramBot/assets/7125622/ef9355f9-55ee-419e-b4b8-94d6b0c99281">

Gaileks committed:
Сократил количество классов
Сократил количество вызовов (пересоздания) словаря.
Сократил количество  MutableList

danelyan reviewed (approved):
Давай перейдем к следующим этапам. Думаю когда картина будет полнее, будет проще  работать над архитектурой )
